lib/model/content.dart

```diff
@@ -1003,22 +1015,45 @@ class _ZulipContentParser {
         continue;
       }
       if (currentParagraph.isNotEmpty) consumeParagraph();
-      result.add(parseBlockContent(node));
+      final block = parseBlockContent(node);
+      if (block is ImageNode) {
+        imageNodes.add(block);
+        continue;
+      }
+      if (imageNodes.isNotEmpty) {
+        result.add(ImageNodeList(imageNodes));
+        imageNodes = [];
+      }
+      result.add(block);
     }
     if (currentParagraph.isNotEmpty) consumeParagraph();
+    if (imageNodes.isNotEmpty) result.add(ImageNodeList(imageNodes));
 
     return result;
   }
 
   List<BlockContentNode> parseBlockContentList(dom.NodeList nodes) {
     assert(_debugParserContext == _ParserContext.block);
-    final acceptedNodes = nodes.where((node) {
+    final List<BlockContentNode> blocks = [];
```

Comment:
nit:
```suggestion
    final List<BlockContentNode> result = [];
```

That matches the name of the corresponding local in `parseImplicitParagraphBlockContentList`. The two functions are very similar now (in fact I should probably follow up by refactoring them into one); so it's best to avoid introducing unnecessary differences, in order to help the real differences stand out when comparing them.